### PR TITLE
IPFS Upload + Delete

### PR DESCRIPTION
**Pull request by valentinyanakiev**
To test upload:

```bash
actionUrl=$(\
    curl -s -X GET -H "Accept: application/json" \
    "http://localhost:3000/identity/ory/kratos/public/self-service/login/api" \
    | jq -r '.ui.action'\
    )
sessionToken=$(\
curl -s -X POST -H  "Accept: application/json" -H "Content-Type: application/json" \
    -d '{"password_identifier": "admin@alkem.io", "password": "your_password", "method": "password"}' \
    "$actionUrl" | jq -r '.session_token' \
    )
```

You can test (assuming default endpoint configuration) creating a file and then uploading it with the following CURL request:

```bash
touch hello-alkemio.pdf

curl http://localhost:3000/api/private/non-interactive/graphql \
 -H "Authorization: Bearer $sessionToken" \
 -F operations='{"query":"mutation UploadFile($file:Upload!) {uploadFile(file:$file)}", "variables": { "file": null }}' \
 -F map='{ "0": ["variables.file"] }' \
 -F 0=@"hello-alkemio.pdf"
```

Deletion is actual [unpinning](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/PIN.md#ipfspinrmipfspath-options) in IPFS + forced garbage collection. Removal of a file is in the graphql-samples, e.g.:

```gql
mutation deleteFile($deleteData: DeleteFileInput!) {
  deleteFile(deleteData: $deleteData)
}
```

```json
{
  "deleteData":
  {
    "CID": "Qmde6CnXDGGe7Dynz1pnxgNARtdVBme9YBwNbo4HJiRy2W"
  }
}

```

After unpinning a garbage collection is enforced and triggered (note: it may take some time to execute GC + do a hard reload on your browser to validate when a file is removed as files, including IPFS ones, are often in cache...).

Supported upload fileS:
- bmp
- jpeg
- x-png
- png
- gif
- pdf

NOTE: to support files e.g. Excel or Word will be an extensive list. Usually different versions of these applications have different mime types.

NOTE2: IPFS has different APIs for its mutable file system (MFS) and the IPFS file system. There are IPFS paths (e.g. `/ipfs/Qmde6CnXDGGe7Dynz1pnxgNARtdVBme9YBwNbo4HJiRy2W` and MFS paths (e.g. `/tmp/hello-alkemio.pdf`) and these are not compatible.

### CODE REVIEW

File: src/domain/common/file-manager/file.manager.service.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    fileName: string,
+    mimetype: string
+  ): Promise<string> {
+    if (!readStream)
+      throw new ValidationException(
+        'Readstream should be defined!',
+        LogContext.FILE_MANAGER
+      );
+
+    if (!mimetype)
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]: `readStream` and `mimetype` are not nullable to begin with.

File: src/domain/common/file-manager/file.manager.service.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+      return true;
+    } catch (error: any) {
+      throw new IpfsDeleteFailedException(
+        `Ipfs removing file at path ${CID} failed! Error: ${error.message}`
+      );
+    }
+  }
+
+  private async validateMimeTypes(mimeType: string): Promise<boolean> {
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]: this function is synchronous

File: src/domain/common/file-manager/file.manager.service.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    const mimeTypes: string = this.configService.get(ConfigurationTypes.STORAGE)
+      ?.ipfs.mime_types;
+    const allowedMimeTypes: string[] = mimeTypes.split(',');
+    if (!allowedMimeTypes.includes(mimeType)) return false;
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]:
```javascript
return allowedMimeTypes.includes(mimeType)
```

File: src/domain/common/file-manager/file.manager.service.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        `Tried to import invalid mimetype ${mimetype}!`,
+        LogContext.FILE_MANAGER
+      );
+    }
+
+    const buffer = await streamToBuffer(readStream);
+
+    try {
+      return await this.ipfsService.uploadFileFromBuffer(buffer);
+    } catch (error: any) {
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]: The correct type is `unknown` here. 
Try something like the below.  [More on the topic](https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript)
 ```javascript
(error as Error).message ?? String(error)
```

File: src/services/adapters/ipfs/ipfs.service.ts [hunk 5]
```diff
@@ -1,29 +1,30 @@
...
+import { CID, create, IPFSHTTPClient } from 'ipfs-http-client';
+import { IpfsDeleteFailedException } from '@common/exceptions/ipfs.delete.exception';
+import { IpfsGCFailedException } from '@common/exceptions/ipfs.gc.exception';
 
 @Injectable()
 export class IpfsService {
-  private ipfsEndpoint = '';
-  private ipfsClientEndpoint = '';
+  private ipfsClient: IPFSHTTPClient;
+  private ipfsClientEndpoint;
```
Comment (hero101) [inline src/services/adapters/ipfs/ipfs.service.ts]: Fields assigned only in the constructor can be made `readonly`.

File: src/services/adapters/ipfs/ipfs.service.ts [hunk 6]
```diff
@@ -41,13 +42,47 @@ export class IpfsService {
...
+        );
+
+        if (gcFile.err) {
+          this.logger.error(
+            `Error in collection ${gcFile.err}`,
+            LogContext.IPFS
+          );
+        }
+      }
+    } catch (error: any) {
```
Comment (hero101) [inline src/services/adapters/ipfs/ipfs.service.ts:80]: same as the other related comment

File: src/platform/authorization/platform.authorization.service.ts [hunk 7]
```diff
@@ -129,11 +132,18 @@ export class PlatformAuthorizationService {
...
       AuthorizationPrivilege.CREATE
     );
     privilegeRules.push(createPrivilege);
 
+    const deletePrivilege = new AuthorizationPolicyRulePrivilege(
+      [AuthorizationPrivilege.FILE_DELETE],
+      AuthorizationPrivilege.DELETE
+    );
+    privilegeRules.push(deletePrivilege);
+
```
Comment (hero101) [inline src/platform/authorization/platform.authorization.service.ts:146]: Could you explain how is this tide up to a specific role?
I don't believe the privileges are assigned correctly as per the AC.

File: src/domain/common/file-manager/file.manager.service.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      );
+    }
+  }
+
+  async removeFile(CID: string): Promise<boolean> {
+    try {
+      await this.ipfsService.unpinFile(CID);
+      await this.ipfsService.garbageCollect();
+
+      return true;
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]: return the success outside of the try block

File: src/domain/common/file-manager/file.manager.resolver.mutations.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      `file upload: ${filename}`
+    );
+    const readStream = createReadStream();
+    const uri = await this.fileManagerService.uploadFile(
+      readStream,
+      filename,
+      mimetype
+    );
+
+    return uri;
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.resolver.mutations.ts]: return the method directly

File: src/domain/common/file-manager/file.manager.resolver.mutations.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      this.platformAuthorizationService.getPlatformAuthorizationPolicy();
+
+    await this.authorizationService.grantAccessOrFail(
+      agentInfo,
+      authorizationPolicy,
+      AuthorizationPrivilege.FILE_DELETE,
+      `file delete: ${CID}`
+    );
+
+    return await this.fileManagerService.removeFile(CID);
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.resolver.mutations.ts]: `await` is not needed

File: src/domain/common/file-manager/file.manager.resolver.mutations.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    );
+
+    return uri;
+  }
+
+  @UseGuards(GraphqlGuard)
+  @Mutation(() => Boolean, {
+    description: 'Removes a file.',
+  })
+  async deleteFile(
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.resolver.mutations.ts]: missing method modifiers as per the SEE agreement

File: src/domain/common/file-manager/file.manager.service.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    try {
+      return await this.ipfsService.uploadFileFromBuffer(buffer);
+    } catch (error: any) {
+      throw new IpfsUploadFailedException(
+        `Ipfs upload of ${fileName} failed! Error: ${error.message}`
+      );
+    }
+  }
+
+  async removeFile(CID: string): Promise<boolean> {
```
Comment (hero101) [inline src/domain/common/file-manager/file.manager.service.ts]: missing method modifiers as per the SEE agreement

**Review by hero101 — CHANGES_REQUESTED** [review]
- Move the ipfs exceptions to an ipfs folder, to improve visibility
- I believe we should create a scalar type for the CID and validate it via a package like [s-ipfs](https://github.com/ipfs-shipyard/is-ipfs) for example
- the `max_file_size` var in `alkemio.yml` is 1 MB - and if reused, must be 5MB per the AC

File: src/domain/common/file-manager/file.manager.service.ts (hunk 1 above)
Comment (valentinyanakiev) [reply src/domain/common/file-manager/file.manager.service.ts]: I agree about the CID, will have a look. Also thought about the ipfs exceptions and it makes sense, but it should be applied structurally as a pattern for all types, and we haven't done it in the past, hence I didn't do it. I will have a look at only it and what would make sense for the rest (won't have refactoring more items in this PR thou, just addressing the comment). The file size I haven't changed but I think your comment is redundant as this is development configuration and the prod configuration is provided in a variable that we can change / manage separately.

File: src/platform/authorization/platform.authorization.service.ts (hunk 7 above)
Comment (valentinyanakiev) [reply src/platform/authorization/platform.authorization.service.ts:146]: There were no acceptance criteria for deletion at all, I have discussed this with Neil and taken this approach. Permissions in this space will need to tidy up regardless as the AC was not granularly defined.

**Pull request merged**